src/group_theory/nielsen_schreier.lean

```diff
@@ -0,0 +1,334 @@
+  of := λ e, loop_of_hom T (of e.val.arrow),
+  unique_lift := begin
+    introsI X _ f,
+    let f' : (generators : quiver G).labelling X := λ a b e,
+      if h : sum.inl e ∈ T a b ∨ sum.inr e ∈ T b a then 1
+      else f ⟨⟨a, b, e⟩, h⟩,
+    rcases unique_lift f' with ⟨F', hF', uF'⟩,
+    refine ⟨F'.map_End _, _, _⟩,
+    { suffices : ∀ {x y} (q : x ⟶ y), F'.map (loop_of_hom T q) = (F'.map q : X),
+      { rintro ⟨⟨a, b, e⟩, h⟩,
+        rw [F'.map_End_apply T♯.root, this, hF'],
+        exact dif_neg h },
+      intros,
+      suffices : ∀ {a} (p : T♯.path T♯.root a), F'.map (hom_of_path T p) = 1,
+      { simp only [this, tree_hom, single_obj.comp_as_mul, single_obj.inv_as_inv, loop_of_hom,
+        one_inv, mul_one, one_mul, functor.map_inv, functor.map_comp] },
+      intros a p, induction p with b c p e ih,
+      { rw [hom_of_path, F'.map_id, single_obj.id_as_one] },
+      rcases e with ⟨e | e, eT⟩,
+      { rw [hom_of_path, F'.map_comp, single_obj.comp_as_mul, ih, mul_one, hF'],
+        exact dif_pos (or.inl eT) },
+      { rw [hom_of_path, F'.map_comp, single_obj.comp_as_mul, ih, mul_one, F'.map_inv,
+          single_obj.inv_as_inv, inv_eq_one, hF'],
+        exact dif_pos (or.inr eT) } },
+    { intros E hE,
+      ext,
+      suffices : (functor_of_monoid_hom T E).map x = F'.map x,
+      { simpa only [loop_of_hom, functor_of_monoid_hom_map, is_iso.inv_id, tree_hom_root,
+          category.id_comp, category.comp_id] using this },
+      congr,
+      apply uF',
+      intros a b e,
+      change E (loop_of_hom T _) = dite _ _ _,
+      split_ifs,
+      { rw [loop_of_hom_eq_id T h, ←End.one_def, E.map_one] },
+      { exact hE ⟨⟨a, b, e⟩, h⟩ } }
+  end }
+
+end retract
+
+open is_free_groupoid quotient_group quiver
+
+instance generators_connected (G) [groupoid.{u u} G] [is_connected G] [is_free_groupoid G] (r : G) :
+  (generators : quiver G).symmetrify.rooted_connected r :=
+begin
+  refine ⟨λ b, _⟩,
+  rw [←quiver.weakly_connected_component.eq, ←free_group.of_injective.eq_iff, ←mul_inv_eq_one],
+  rcases category_theory.nonempty_hom_of_connected_groupoid b r with ⟨p⟩,
+  let X := free_group (generators : quiver G).weakly_connected_component,
+  let f : G → X := λ g, free_group.of ↑g,
+  let F : G ⥤ single_obj X := single_obj.difference_functor f,
+  change F.map p = ((category_theory.functor.const G).obj ()).map p,
+  congr, ext,
+  rw [functor.const.obj_map, single_obj.id_as_one,
+    single_obj.difference_functor_map, mul_inv_eq_one],
+  apply congr_arg free_group.of,
+  rw quiver.weakly_connected_component.eq,
+  exact ⟨quiver.arrow.to_path (sum.inr e)⟩,
+end
+
+instance End_is_free_group_of_connected_free {G} [groupoid G] [is_connected G] [is_free_groupoid G]
+  (r : G) : is_free_group (End r) :=
+End_is_free_group_of_arborescence (quiver.geodesic_subtree _ r)
+
+instance subgroup_is_free_of_free {G} [group.{u} G] [is_free_group G]
+  (H : subgroup G) : is_free_group H :=
```

Comment:
This `instance` should have a doc-string containing the string "Nielsen-Schreier".